Dataflow integration test

cisaacstern opened the pull request:
WIP closes #51 when complete.

So far, I've run this test locally and confirmed that it deploys a dataflow job which succeeds. The job takes about 6 minutes to complete.

Next steps:

- [x] Add some assertions in the test regarding the completed job (that it succeeded, perhaps that it built data)
- [x] Add GitHub Workflow for running this test from GitHub
- [x] Demonstrate that this test fails without downgrading `pangeo-forge-recipes` version and `sdk_container_version` tag
- [x] Downgrade this version + tag and re-run test, demonstrating that it passes

cisaacstern commented:
The current test passes for me locally with output

<details>

<summary>pytest output</summary>

```console
$ pytest -vvvxs tests/test_dataflow_integration.py
===================================================== test session starts ======================================================
platform darwin -- Python 3.9.15, pytest-7.2.1, pluggy-1.0.0 -- /Users/charlesstern/miniconda3/envs/pfr/bin/python3.9
cachedir: .pytest_cache
rootdir: /Users/charlesstern/Dropbox/pangeo/pangeo-forge-runner
plugins: mock-3.10.0, cov-4.0.0
collected 1 item   
                                                                                                       
tests/test_dataflow_integration.py::test_dataflow_integration
Submitting job...
Job submitted with job_id = '2023-01-18_16_02_51-17714952298081130589'
Waiting for 5 mins, starting at start = 1674086572.854805
Time elapsed = 300.00544595718384
Current state = 'Running'
Time elapsed = 331.8521089553833
Current state = 'Running'
Time elapsed = 363.6149618625641
Current state = 'Done'
<xarray.Dataset>
Dimensions:     (latitude: 180, nv: 2, longitude: 360, time: 2)
Coordinates:
    lat_bounds  (latitude, nv) float32 ...
  * latitude    (latitude) float32 -90.0 -89.0 -88.0 -87.0 ... 87.0 88.0 89.0
    lon_bounds  (longitude, nv) float32 ...
  * longitude   (longitude) float32 0.0 1.0 2.0 3.0 ... 356.0 357.0 358.0 359.0
  * time        (time) datetime64[ns] 1996-10-01 1996-10-02
Dimensions without coordinates: nv
Data variables:
    precip      (time, latitude, longitude) float32 ...
Attributes: (12/48)
    Conventions:                CF-1.6, ACDD 1.3
    Metadata_Conventions:       CF-1.6, Unidata Dataset Discovery v1.0, NOAA ...
    acknowledgment:             This project was supported in part by a grant...
    cdm_data_type:              Grid
    cdr_program:                NOAA Climate Data Record Program for satellit...
    cdr_variable:               precipitation
    ...                         ...
    standard_name_vocabulary:   CF Standard Name Table (v41, 22 February 2017)
    summary:                    Global Precipitation Climatology Project (GPC...
    time_coverage_duration:     P1D
    time_coverage_end:          1996-10-01T23:59:59Z
    time_coverage_start:        1996-10-01T00:00:00Z
    title:                      Global Precipitation Climatatology Project (G...
PASSED

================================================ 1 passed in 400.14s (0:06:40) =================================================
```

</details>

with local changes to pangeo-forge-runner (pre-installation)

<details>

<summary>diff</summary>

```diff
diff --git a/pangeo_forge_runner/commands/bake.py b/pangeo_forge_runner/commands/bake.py
index 3509865..cd73218 100644
--- a/pangeo_forge_runner/commands/bake.py
+++ b/pangeo_forge_runner/commands/bake.py
@@ -77,8 +77,8 @@ class Bake(BaseCommand):
     )
 
     container_image = Unicode(
-        # Provides apache_beam 2.43, which we pin to in setup.py
-        "quay.io/pangeo/forge:8904ade",
+        # Provides apache_beam 2.42, which we pin to in setup.py
+        "quay.io/pangeo/forge:5e51a29",
         config=True,
         help="""
         Container image to use for this job.
diff --git a/setup.py b/setup.py
index 9ec3c7c..c748350 100644
--- a/setup.py
+++ b/setup.py
@@ -16,11 +16,11 @@ setup(
     install_requires=[
         "jupyter-repo2docker",
         "ruamel.yaml",
-        "pangeo-forge-recipes>=0.9.4",
+        "pangeo-forge-recipes==0.9.2",
         "traitlets",
         # Matches the version of apache_beam in the default image,
         # specified in bake.py's container_image traitlet default
-        "apache-beam[gcp]==2.43.0",
+        "apache-beam[gcp]==2.42.0",
     ],
     entry_points={
         "console_scripts": ["pangeo-forge-runner=pangeo_forge_runner.cli:main"]
```

</details>

Next I'll work on getting this same result in GitHub Actions.

yuvipanda commented on .github/workflows/unit-test.yml:
```diff
@@ -45,7 +45,10 @@ jobs:
       run: |
         # Test everything except the flink tests
-        pytest -vvv -s --cov=pangeo_forge_runner --ignore=tests/test_flink.py tests/
+        pytest -vvv -s --cov=pangeo_forge_runner \
```
Let's split the tests into a 'unit/' directory and an 'integration/' directory, so this can just run unit tests instead of having to explicitly add each integration test as an --ignore?

cisaacstern replied:
Done in [a6da8c0](https://github.com/pangeo-forge/pangeo-forge-runner/pull/52/commits/a6da8c061f9a8dad4caf55f0b9a4547fb3e93799)

cisaacstern commented:
@yuvipanda, notes on this PR:

- For GCP auth, I've added a GCP service account key to this repo's secrets as `GOOGLE_CREDENTIALS`. Perhaps we'll want to upgrade to the recommended Workload Identity Federation eventually, but the [setup for that seemed pretty involved](https://github.com/google-github-actions/auth#setting-up-workload-identity-federation), so I'm just using the keyfile approach now to move fast.
- The **Dataflow Integration Test** runs on all pushes to `main`. For PRs, it runs on commits to the PR _**only if the `test-dataflow` tag is present**_. This can save us deploying a lot of Dataflow jobs for PRs that have nothing to do with Dataflow. Also, I'm pleased with the fact that it will run _**when the `test-dataflow` label is created**_ as well, so if we want to run the test for a PR which did not previously have this label, we can add the label and the job will run without needing a new commit.

    > I've given this a fair amount of thought, and deliberately chosen this label-based dispatch strategy over chatops / slash command based dispatch, because it avoids the pitfall of `issue_comment` events wherein the workflow which responds to these events _must be in the repo's default branch_. I personally find this quite a large barrier to development because it means the PR which develops the workflow cannot also run the workflow (without first adding a dispatching workflow to `main`). 


- [This is the first job](https://github.com/pangeo-forge/pangeo-forge-runner/actions/runs/3962323695/jobs/6788843216) on this PR that actually deployed a job to Dataflow. Note that it timed out (failed) after 12 minutes (a value I've chosen, see comment in test). This timeout strategy seems to work to save time, because this job really did fail, but "official failure" took an hour:

    <img width="1212" alt="Screen Shot 2023-01-19 at 2 34 08 PM" src="https://user-images.githubusercontent.com/62192187/213577350-c3d9c8a2-bb35-4b3a-bdc8-8b94f2211504.png">

- The **Dataflow Integration Test** which ran from https://github.com/pangeo-forge/pangeo-forge-runner/pull/52/commits/8c2f353c89fa1363692b5dfdf230dc3b2681a599 succeeded, demonstrating that my PR https://github.com/pangeo-forge/pangeo-forge-runner/pull/50 added an incompatible default `sdk_container_image` tag into `main`.

yuvipanda commented on .github/workflows/dataflow.yaml:
```diff
@@ -0,0 +1,48 @@
+    - name: 'Authenticate to Google Cloud'
+      uses: 'google-github-actions/auth@v1'
+      with:
+        credentials_json: '${{ secrets.GOOGLE_CREDENTIALS }}'
```
Can we make this something more specific, like `GCP_DATAFLOW_SERVICE_KEY`?

cisaacstern replied:
Good call. Fixed in 5523840.

yuvipanda commented:
One comment about naming, happy to merge after that, @cisaacstern :)

Another suggestion I have is to also have the integration test run on a [schedule](https://docs.github.com/en/actions/using-workflows/events-that-trigger-workflows#schedule), to catch any changes that might happen in dataflow itself.

Also, how do we make sure we notice the test failure itself?

cisaacstern commented:
Thanks for the review, @yuvipanda.

Changed the credential name and added a nightly scheduled run (both great calls).

> Also, how do we make sure we notice the test failure itself?

Could add a branch protection requiring it to pass for PRs? For the nightly run, I will get an email if it fails.

yuvipanda commented:
Thanks a lot for this, @cisaacstern :) I really like the idea of using it with labels!

One other question I have is should we delete the GCS buckets as a form of cleanup after?

cisaacstern commented:
> One other question I have is should we delete the GCS buckets as a form of cleanup after?

As opposed to just letting this bucket grow and grow into infinity? Yeah, probably! 😆 

Maybe this is lazy (as opposed to putting it in code), but I just added a rule to the bucket to delete objects after 3 days:

<img width="504" alt="Screen Shot 2023-01-20 at 3 35 14 PM" src="https://user-images.githubusercontent.com/62192187/213823755-ee9c9061-5b65-436d-af84-14d0882c37ca.png">


I think I like this (vs. deleting on test cleanup) because it gives a little time to look at the objects as part of debugging a test failure. 🤷

yuvipanda commented:
@cisaacstern oh delete after 3 days is actually perfect haha let's keep that. I'd suggest 7days over 3.

cisaacstern commented:
<img width="300" alt="Screen Shot 2023-01-20 at 3 54 39 PM" src="https://user-images.githubusercontent.com/62192187/213825299-cb819b03-8119-417c-9b10-e3f46ce6b339.png">

Done! 👍
